### DISCUSSION TITLE
Limited Reset Options Showing for MACHINE on PICO W

### DISCUSSION THREAD

**Discussion by DS256** [micropython/micropython#10858] [RP2040 / Pico] [answered]
I'm just starting to work with micropython on my PICO W. I've seen the various options for reset in in the [documentation](https://docs.micropython.org/en/latest/library/machine.html). Here's what I see for MACHINE

`>>> help (machine)
object <module 'umachine'> is of type module
  __name__ -- umachine
  unique_id -- <function>
  soft_reset -- <function>
  reset -- <function>
  reset_cause -- <function>
  bootloader -- <function>
  freq -- <function>
  idle -- <function>
  lightsleep -- <function>
  deepsleep -- <function>
  disable_irq -- <function>
  enable_irq -- <function>
  bitstream -- <function>
  time_pulse_us -- <function>
  dht_readinto -- <function>
  mem8 -- <8-bit memory>
  mem16 -- <16-bit memory>
  mem32 -- <32-bit memory>
  ADC -- <class 'ADC'>
  I2C -- <class 'I2C'>
  SoftI2C -- <class 'SoftI2C'>
  I2S -- <class 'I2S'>
  Pin -- <class 'Pin'>
  PWM -- <class 'PWM'>
  RTC -- <class 'RTC'>
  Signal -- <class 'Signal'>
  SPI -- <class 'SPI'>
  SoftSPI -- <class 'SoftSPI'>
  Timer -- <class 'Timer'>
  UART -- <class 'UART'>
  WDT -- <class 'WDT'>
  PWRON_RESET -- 1
  WDT_RESET -- 3`

When I power cycle, I get the reset_cause of 1 (PWRON_RESET) reset() and soft_reset() return 3 (WDT_RESET).

I'm using [Pico Display](https://shop.pimoroni.com/products/pico-display-pack) and loaded [their firmware](https://github.com/pimoroni/pimoroni-pico/releases)  pimoroni-picow-v1.19.12-micropython.uf2 and was wondering if that could be the issue.

My question is this what I'm seeing in help(machine) normal micropython for PICO W? 

Thanks

**Comment by scruss** [comment]
This is what I get on a Pico W:

```
MicroPython v1.19.1-887-gb11026689 on 2023-02-21; Raspberry Pi Pico W with RP2040
Type "help()" for more information.
>>> help(machine)
object <module 'umachine'> is of type module
  __name__ -- umachine
  unique_id -- <function>
  soft_reset -- <function>
  reset -- <function>
  reset_cause -- <function>
  bootloader -- <function>
  freq -- <function>
  idle -- <function>
  lightsleep -- <function>
  deepsleep -- <function>
  disable_irq -- <function>
  enable_irq -- <function>
  bitstream -- <function>
  time_pulse_us -- <function>
  dht_readinto -- <function>
  mem8 -- <8-bit memory>
  mem16 -- <16-bit memory>
  mem32 -- <32-bit memory>
  ADC -- <class 'ADC'>
  I2C -- <class 'I2C'>
  SoftI2C -- <class 'SoftI2C'>
  I2S -- <class 'I2S'>
  Pin -- <class 'Pin'>
  PWM -- <class 'PWM'>
  RTC -- <class 'RTC'>
  Signal -- <class 'Signal'>
  SPI -- <class 'SPI'>
  SoftSPI -- <class 'SoftSPI'>
  Timer -- <class 'Timer'>
  UART -- <class 'UART'>
  WDT -- <class 'WDT'>
  PWRON_RESET -- 1
  WDT_RESET -- 3
```

MicroPython help text is usually extremely terse. Best to the current (*not* ‘latest’) [Reset related functions](https://docs.micropython.org/en/v1.19.1/library/machine.html#reset-related-functions) online.

**Reply by DS256** [reply]
Thanks. Looks like you get the same results as I do.

I had checked your link and the reset causes were listed as
machine.PWRON_RESET¶
machine.HARD_RESET¶
machine.WDT_RESET¶
machine.DEEPSLEEP_RESET¶
machine.SOFT_RESET¶

**Reply by DS256** [reply]
This discussion expanded to issues with MAIN not running under all resets which seems to be associated when Thonny is attached. See those details and confirmation on what I have been able to get working at https://github.com/orgs/micropython/discussions/10884

**Comment by DS256** [comment, marked as answer]
See https://github.com/orgs/micropython/discussions/10884

**Comment by br4cine** [comment]
Try this:

reset_source = machine.mem32[0x40058008].
You'll get 0 for normal power on. 1 for watchdog and 2 for machine.reset.

**Comment by kjm1102** [comment]
Be aware some chips on some IDEs will change their claimed pwrup_cause depending on if your program is running from within the chip (eg as boot.py) or if you run it from your PC via the IDE (typically during development). I'm wishing now I'd made a note of which chip & which IDE was telling me porkies  so I could test br4cine's mem32 location for invariance.
